cdap-security/src/main/java/co/cask/cdap/security/impersonation/DefaultUGIProvider.java

```diff
@@ -63,8 +79,42 @@
    * @throws IOException if there was any IOException during localization of the keytab
    */
   @Override
-  protected UserGroupInformation createUGI(ImpersonationInfo impersonationInfo) throws IOException {
-    LOG.debug("Configured impersonation info: {}", impersonationInfo);
+  protected UGIWithPrincipal createUGI(ImpersonationOpInfo impersonationOpInfo) throws IOException {
+    if (impersonationOpInfo.getEntityId().getEntityType().equals(EntityType.NAMESPACE)) {
+      // CDAP-8355 If the operation being impersonated is an explore query then check if the namespace configuration
+      // specifies that it can be impersonated with the namespace owner.
+      // This is done here rather than in the get getConfiguredUGI because the getConfiguredUGI will be called at
+      // remote location as in RemoteUGIProvider. Looking up namespace meta there will make a trip to master followed
+      // by one to get the credentials. Hence do it here in the DefaultUGIProvider which is running in master.
+      // Although, this will cause cache miss for explore implementation if the namespace config doesn't allow
+      // impersonating the namespace owner for explore queries but that a trade-off to avoid multiple remote calls in
+      // more prominent calls.
+      try {
+        NamespaceConfig nsConfig = namespaceQueryAdmin.get(impersonationOpInfo.getEntityId().getNamespaceId())
+          .getConfig();
+        if (impersonationOpInfo.getImpersonatedOpType().equals(ImpersonatedOpType.EXPLORE) &&
```

Comment:
perhaps you can check whether this is an explore query before retrieving the namespace meta? That saves you a read if it is not for explore.